Adds VirtualBox provisioning information (#318)
coreos/fedora-coreos-docs: modules/ROOT/pages/provisioning-virtualbox.adoc

= Provisioning Fedora CoreOS on VirtualBox

This guide shows how to provision new Fedora CoreOS (FCOS) nodes on the VirtualBox hypervisor.

== Prerequisites

Before provisioning an FCOS machine, you must have an Ignition configuration file containing your customizations and host it somewhere (e.g. using `python3 -m http.server`, we will assume this option is used in the following examples). If you do not alrady have an Ignition file, see xref:producing-ign.adoc[Producing an Ignition File].

== Setting up a new VM

> You can do set up VirtualBox based Virtual Machine through the GUI or via the https://www.virtualbox.org/manual/UserManual.html#vboxmanage[`vboxmanage` CLI]. Below is a shell script that does most of what the graphical wizard will do, however, is tailored to the installation from live ISO or Raw disk image as outlined in the following sections.

Reviewer: ```suggestion
You can set up a VirtualBox based Virtual Machine through the GUI or via the https://www.virtualbox.org/manual/UserManual.html#vboxmanage[`vboxmanage` CLI]. Below is a shell script that does most of what the graphical wizard will do. Note that it is tailored to the installation from live ISO or Raw disk image as outlined in the following sections.
```

Author: Fixed

    quay.io/coreos/coreos-installer:release download -f iso
----

Before the installation can begin you have to convert the raw disk image to a VirtualBox compatible one:
[source, bash]
----
FCOS_DISK=$(realpath ~/Downloads/fedora-coreos-34.20210808.3.0-metal.x86_64.raw.xz)
UNPACKED_DISK=$(realpath ~/Downloads/fedora-coreos-34.20210808.3.0-metal.x86_64.raw)

# Unpack the raw disk image
unxz "${FCOS_DISK}"
----

[NOTE]
====
If you want to supply the Ignition configuration file URL statically, you can modify the raw disk before converting it to a VirtualBox disk image:
[source, bash]
----
TMP_FCOS_BOOT_VOLUME=$(realpath /tmp/fcos_boot_volume)
mkdir "${TMP_FCOS_BOOT_VOLUME}"

fdisk -l "${UNPACKED_DISK}"
# calculate the $OFFSET for the mount operation based on the sector size of the disk and the start point of the boot partion
sudo mount -o loop,offset="${OFFSET}" fedora-coreos-34.20210808.3.0-metal.x86_64.raw "${TMP_FCOS_BOOT_VOLUME}"
# the boot config you are looking for is located in /loader/entries/ostree-1-fedora-coreos.conf on the boot partition
sudo vi "${TMP_FCOS_BOOT_VOLUME}/loader/entries/ostree-1-fedora-coreos.conf"
# ...
sudo umount "${TMP_FCOS_BOOT_VOLUME}"
----

> The file contains a line containing the string `ignition.platform.id=metal $ignition_firstboot` – add `ignition.config.url=http://10.0.2.2:8000/example.ign` after the first boot parameter and save it. You can now unmount the disk.

Reviewer: Maybe split the code block in two and move this line right under the vi command?

Author: I split that note up a little, please check whether that improves readability.

= Provisioning Fedora CoreOS on VirtualBox

> This guide shows how to provision new Fedora CoreOS (FCOS) nodes on the VirtualBox hypervisor.

Reviewer 2: It feels like we should right upfront mention that FCOS currently does not ship VirtualBox images but may in the future and that what follows are platform-agnostic ways to set up Fedora CoreOS as a bare metal system.

Author: That's a good point. Would it be alright to add that hint together with links to the two issues ([#144](https://github.com/coreos/fedora-coreos-tracker/issues/144), [#73](https://github.com/coreos/fedora-coreos-tracker/issues/73)) related to providing VirtualBox images?

Reviewer 2: Yes, that sounds great!

Author: I updated the pull request with such a hint.

----

You can now boot the target system you have configured. The ISO is capable of bringing up a fully functioning FCOS system purely from memory (i.e. without using any disk storage). Once booted, you will have access to a bash command prompt.

You can now run `coreos-installer` inside the VM:
[source, bash]
----
sudo coreos-installer install /dev/sda \
    --insecure-ignition --ignition-url http://10.0.2.2:8000/example.ign
----

TIP: Check out `coreos-installer install --help` for more options on how to install Fedora CoreOS.

Once the installation is complete, you have to remove the live ISO after shutting down (`sudo shutdown -h now`) the VM. Once the ISO is removed, you can start the VM anew. The actual first boot process begins now. It is at this time that Ignition ingests the configuration file and provisions the system as specified.

[source, bash]
----
vboxmanage storageattach "${VM_NAME}" --storagectl IDE --type dvddrive --port 1 --device 0 --medium emptydrive
vboxmanage startvm "${VM_NAME}" --type headless
# Once the VM is available, you can log in:
ssh core@localhost -p 2222
----

> == Installing from Raw

Reviewer 2: I appreciate the work that went behind writing and testing this. But personally I'm hesitant to document it as if it were a supported workflow. I think to start, just having the ISO install path would be great. Or otherwise if folks really want, maybe we should at least have a disclaimer around here that this is not heavily tested and may break.

Author: That somewhat mirrors my own thoughts, in particular concerning the actual locations of the config file, disk name etc. At the same time this process works spectacularly well (imho better than the regular ISO install with the cumbersome "press e at the right time" issue) if the user is already on Linux. Also, technically I should've probably used `parted` instead of `fdisk` but in the end it probably doesn't matter too much.
I am definitely in favor of the disclaimer, though.

Reviewer: Let's make a big warning and keep it? It's also already mentioned in a second time after the ISO. I understand it's a bit fragile but it's a nice performance improvement for non casual users.

Otherwise we could split this into 2 PRs to merge the first part and make a new PR to discuss improvements for the raw disk part.

Reviewer 2: Yup sure, I'm OK with keeping it and just adding a warning.

Author: I updated the pull request with such a warning.

Reviewer 2: ```suggestion
== Installing from raw bare metal image
```

= Provisioning Fedora CoreOS on VirtualBox

This guide shows how to provision new Fedora CoreOS (FCOS) nodes on the VirtualBox hypervisor.

== Prerequisites

> Before provisioning an FCOS machine, you must have an Ignition configuration file containing your customizations and host it somewhere (e.g. using `python3 -m http.server`, we will assume this option is used in the following examples). If you do not alrady have an Ignition file, see xref:producing-ign.adoc[Producing an Ignition File].

Reviewer 2: ```suggestion
Before provisioning an FCOS machine, you must have an Ignition configuration file containing your customizations and host it somewhere (e.g. using `python3 -m http.server`, we will assume this option is used in the following examples). If you do not already have an Ignition file, see xref:producing-ign.adoc[Producing an Ignition File].
```

== Setting up a new VM

> You can set up a VirtualBox based Virtual Machine through the GUI or via the https://www.virtualbox.org/manual/UserManual.html#vboxmanage[`vboxmanage` CLI]. Below is a shell script that does most of what the graphical wizard will do. Note that it is tailored to the installation from live ISO or Raw disk image as outlined in the following sections.

Reviewer 2: ```suggestion
You can set up a VirtualBox virtual machine through the GUI or via the https://www.virtualbox.org/manual/UserManual.html#vboxmanage[`vboxmanage` CLI]. Below is a shell script that does most of what the graphical wizard will do. Note that it is tailored to the installation from live ISO or raw disk image as outlined in the following sections.
```

NOTE: Please be aware that for now there is no dedicated Fedora CoreOS image available for VirtualBox and the installation instructions provided are essentially xref:bare-metal.adoc[bare metal instructions]. A dedicated image may be provided at a later time. Please check the GitHub issues https://github.com/coreos/fedora-coreos-tracker/issues/73[#73] and https://github.com/coreos/fedora-coreos-tracker/issues/144[#144] for more details.

> We will rely on the VirtualBox.provided NAT gateway to access the hosts' loopback interface on `10.0.2.2`. Check the https://www.virtualbox.org/manual/UserManual.html#network_nat[VirtualBox User Manual] for more details.

Reviewer 2: ```suggestion
We will rely on the VirtualBox-provided NAT gateway to access the hosts' loopback interface on `10.0.2.2`. Check the https://www.virtualbox.org/manual/UserManual.html#network_nat[VirtualBox User Manual] for more details.
```

Author: Ooops 😅

> We will rely on the VirtualBox provided NAT gateway to access the hosts's loopback interface on `10.0.2.2`. Check the https://www.virtualbox.org/manual/UserManual.html#network_nat[VirtualBox Documentation] for more details.

Reviewer 2: ```suggestion
We will rely on the VirtualBox-provided NAT gateway to access the hosts' loopback interface on `10.0.2.2`. Check the https://www.virtualbox.org/manual/UserManual.html#network_nat[VirtualBox User Manual] for more details.
```

[source, bash]
----
VM_NAME='fcos-node01'
VM_DISK=$(realpath ~/VirtualBox\ VMs/${VM_NAME})/${VM_NAME}.vdi

vboxmanage createvm --name "${VM_NAME}" --ostype Fedora_64 --register
# Set hardware settings for the VM, including recommended graphics controller

> vboxmanage modifyvm "${VM_NAME}" --memory 4096 --cpus 2 --vram 20 --graphicscontroller vmsvga --rtcuseutc on --usb on --usbehci on

Reviewer 2: Is USB actually required?

Author: Fair question: I doubt it but as described in the leading paragraph, the shell command(s) are designed to give a very similar output the GUI wizard. I can test that again without USB and fix this later accordingly.

----

You can now boot the target system you have configured. The ISO is capable of bringing up a fully functioning FCOS system purely from memory (i.e. without using any disk storage). Once booted, you will have access to a bash command prompt.

You can now run `coreos-installer` inside the VM:
[source, bash]
----
sudo coreos-installer install /dev/sda \
    --insecure-ignition --ignition-url http://10.0.2.2:8000/example.ign
----

TIP: Check out `coreos-installer install --help` for more options on how to install Fedora CoreOS.

Once the installation is complete, you have to remove the live ISO after shutting down (`sudo shutdown -h now`) the VM. Once the ISO is removed, you can start the VM anew. The actual first boot process begins now. It is at this time that Ignition ingests the configuration file and provisions the system as specified.

[source, bash]
----
vboxmanage storageattach "${VM_NAME}" --storagectl IDE --type dvddrive --port 1 --device 0 --medium emptydrive
vboxmanage startvm "${VM_NAME}" --type headless
# Once the VM is available, you can log in:
ssh core@localhost -p 2222
----

> == Installing from Raw
> 

Reviewer 2: ```suggestion
== Installing from raw bare metal image

```

Author: Changed.

WARNING: The steps below are provided for advanced users, may unexpectedly break in various places with newer Fedora CoreOS images, and require a Unix environment to work. If you encounter issues consider using the xref:_installing_from_live_iso[live ISO instructions] instead.

> To install FCOS into a VirtualBox virtual machine using the Raw disk image, follow these steps:

Reviewer 2: ```suggestion
To install FCOS into a VirtualBox virtual machine using the raw disk image, follow these steps:
```
